### DISCUSSION TITLE
Cask as a dependency for Java application

### DISCUSSION THREAD

**Discussion by martinpaljak** [Homebrew/discussions#4792] [Writing Formulae/Casks]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.11
ORIGIN: https://github.com/Homebrew/brew
HEAD: 4afb8e5602f3ecc9edf67a44257d8eceeaa8a108
Last commit: 7 days ago
Core tap JSON: 18 Sep 14:16 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: joe
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 14.0-arm64
CLT: N/A
Xcode: 15.0
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You are using macOS 14.
We do not provide support for this pre-release version.
It is expected behaviour that some formulae will fail to build in this pre-release version.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.
```

### Description of issue

Application depends on JavaFX (thus zulufx cask)

How does one declare a dependency on the cask (brew install --cask zulufx)?

https://stackoverflow.com/questions/35461948/syntax-for-a-homebrew-formula-depending-on-a-cask does not help, nor the documentation section for dependencies.

Once the dependency is set, how can I make the app always run with the dependent JDK? Specifically, this wrapper generator:

```
bin.write_jar_script libexec/jar_file, "jarfile", java_version: "11"
```

Thanks!

**Comment by SMillerDev** [comment]
If this is for `homebrew/core`, note that it will not be accepted: https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae

Otherwise, I what exactly does not work about those suggestions? Does it not trigger an install?

**Comment by martinpaljak** [comment]
Not for core (at least now), having a "personal tap".

```
  depends_on "gradle" => :build
  depends_on "zulufx" => :cask
```

Gives

```
Warning: No available formula with the name "zulufx" (dependency of nfc).
==> Searching for similarly named formulae and casks...
==> Casks
zulu                                                                                            zulufx

To install zulu, run:
  brew install --cask zulu
```

**Reply by SMillerDev** [reply]
I don't think any of those responses suggested that though, because that syntax is reserved for a lifecycle stage. Currently you're indicating it isn't needed for runtime, build time or test time. Only for cask time.

Try `depends_on cask: "zulufx"`

**Reply by martinpaljak** [reply]
`depends_on cask: "zulufx"`

gives:


```
# brew install nfc.rb
Error: nfc: Unsupported special dependency :cask
```

**Reply by Bo98** [reply]
The best you can do is to make a custom requirement with a message telling the user what to do. We don't support automatic installs for various reasons, one example being that Casks can require sudo but `brew install` on a formula is guaranteed to never require sudo. Other reasons include Linux support, how non-Cask installs would be handled given casks are just fancy wrappers around upstream installers, and how to treat (for example) different flavours and versions of OpenJDK (or OpenJFX in your case) that may be installed.

The glibc formula has some examples of custom requirements: https://github.com/Homebrew/homebrew-core/blob/master/Formula/g/glibc.rb. You just need to adjust the `satisfy` condition and messages accordingly. If you're fine with the default message, you could get rid of `def message ... end` entirely and replace it with `cask "zulufx"`, which generates a message to direct the user to run `brew install --cask zulufx`.

**Comment by martinpaljak** [comment]
Thanks for the hints! As the main target audience for brew IMHO is command line users, I can omit the FX requirement and make it depend on "normal" java, and thus avoid casks.

Will also investigate custom requirements.
